Avoid panic_bounds_check in fmt::write.

m-ou-se opened the pull request:
Writing any fmt::Arguments would trigger the inclusion of usize formatting and padding code in the resulting binary, because indexing used in fmt::write would generate code using panic_bounds_check, which prints the index and length.

These bounds checks are not necessary, as fmt::Arguments never contains any out-of-bounds indexes.

This change replaces them with unsafe get_unchecked, to reduce the amount of generated code, which is especially important for embedded targets.

---

Demonstration of the size of and the symbols in a 'hello world' no_std binary:

<details>
<summary>Source code</summary>

```rust
#![feature(lang_items)]
#![feature(start)]
#![no_std]

use core::fmt;
use core::fmt::Write;

#[link(name = "c")]
extern "C" {
    #[allow(improper_ctypes)]
    fn write(fd: i32, s: &str) -> isize;
    fn exit(code: i32) -> !;
}

struct Stdout;

impl fmt::Write for Stdout {
    fn write_str(&mut self, s: &str) -> fmt::Result {
        unsafe { write(1, s) };
        Ok(())
    }
}

#[start]
fn main(_argc: isize, _argv: *const *const u8) -> isize {
    let _ = writeln!(Stdout, "Hello World");
    0
}

#[lang = "eh_personality"]
fn eh_personality() {}

#[panic_handler]
fn panic(_: &core::panic::PanicInfo) -> ! {
    unsafe { exit(1) };
}
```
</details>

Before:
```
   text	   data	    bss	    dec	    hex	filename
   6059	    736	      8	   6803	   1a93	before
```
```
0000000000001e00 T <T as core::any::Any>::type_id
0000000000003dd0 D core::fmt::num::DEC_DIGITS_LUT
0000000000001ce0 T core::fmt::num::imp::<impl core::fmt::Display for u64>::fmt
0000000000001ce0 T core::fmt::num::imp::<impl core::fmt::Display for usize>::fmt
0000000000001370 T core::fmt::write
0000000000001b30 t core::fmt::Formatter::pad_integral::write_prefix
0000000000001660 T core::fmt::Formatter::pad_integral
0000000000001350 T core::ops::function::FnOnce::call_once
0000000000001b80 t core::ptr::drop_in_place
0000000000001120 t core::ptr::drop_in_place
0000000000001c50 t core::iter::adapters::zip::Zip<A,B>::new
0000000000001c90 t core::iter::adapters::zip::Zip<A,B>::new
0000000000001b90 T core::panicking::panic_bounds_check
0000000000001c10 T core::panicking::panic_fmt
0000000000001130 t <&mut W as core::fmt::Write>::write_char
0000000000001200 t <&mut W as core::fmt::Write>::write_fmt
0000000000001250 t <&mut W as core::fmt::Write>::write_str
```

After:
```
   text	   data	    bss	    dec	    hex	filename
   3068	    600	      8	   3676	    e5c	after
```
```
0000000000001360 T core::fmt::write
0000000000001340 T core::ops::function::FnOnce::call_once
0000000000001120 t core::ptr::drop_in_place
0000000000001620 t core::iter::adapters::zip::Zip<A,B>::new
0000000000001660 t core::iter::adapters::zip::Zip<A,B>::new
0000000000001130 t <&mut W as core::fmt::Write>::write_char
0000000000001200 t <&mut W as core::fmt::Write>::write_fmt
0000000000001250 t <&mut W as core::fmt::Write>::write_str
```

rust-highfive commented:
r? @dtolnay

(rust_highfive has picked a reviewer for you, use r? to override)

m-ou-se commented:
Related optimization:

https://github.com/rust-lang/rust/blob/a85e94927622665a9e9022de0d33a890a2e32d43/library/core/src/panicking.rs#L44-L50

That optimzation would not be very effective if even writing the most trivial `fmt::Arguments` pulls in formatting code for usizes, like it does now.

Not sure if this can (or should) be tested for. Checking if some code results in a binary smaller than some threshold would not be a good test. Coming up with a list of symbols that 'should not be there' is also not easy.

m-ou-se commented:
@rustbot modify labels: +A-fmt +I-heavy +libs-impl

m-ou-se commented:
@rustbot modify labels: +A-fmt +I-heavy +T-libs

Mark-Simulacrum commented:
I think it would be good to see if we can add a simple codegen test to assert that there's no panic/bounds check here, since it seems easy for it to sneak back in.

I would also like debug asserts here - the compiler *should* be correct, but I'm not sure we're testing that anywhere else (other than kinda via these bounds checks and such).

m-ou-se commented:
@Mark-Simulacrum sounds good. Will add that when I have time.

@rustbot modify labels: +S-waiting-on-author -S-waiting-on-review

m-ou-se commented:
Added the debug_asserts and a test.

The test is a `run-make` test, because it needs to check the result after linking. A `codegen` or `assembly` test doesn't check the parts that will be pulled in from `core` by the linker.

@rustbot modify labels: -S-waiting-on-author +S-waiting-on-review

dtolnay commented:
src/test/run-make/fmt-write-bloat/main doesn't need to be checked in, right?

Mark-Simulacrum commented on src/test/run-make/fmt-write-bloat/main.rs:
```diff
@@ -0,0 +1,33 @@
+#[start]
+fn main(_: isize, _: *const *const u8) -> isize {
+    let _ = writeln!(Dummy, "Hello World");
+    // CHECK-NOT: panic_bounds_check
```
I think this can be removed now -- presumably it's leftover from a codegen test?

Mark-Simulacrum reviewed (commented):
This seems good to me -- r=me with the binary file removed.

m-ou-se commented:
Oh oops. Accidentally committed some things I shouldn't have. Thanks. Updated.

dtolnay commented:
@bors r=Mark-Simulacrum

<details>
<summary>I confirmed that the new test fails without the changes in the PR.</summary>

```
---- [run-make] run-make/fmt-write-bloat stdout ----

error: make failed
status: exit code: 2
command: "make"
stdout:
------------------------------------------
LD_LIBRARY_PATH="/git/rust/build/x86_64-unknown-linux-gnu/test/run-make/fmt-write-bloat/fmt-write-bloat:/git/rust/build/x86_64-unknown-linux-gnu/stage1/lib:/git/rust/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/git/rust/build/x86_64-unknown-linux-gnu/stage0/lib" '/git/rust/build/x86_64-unknown-linux-gnu/stage1/bin/rustc' --out-dir /git/rust/build/x86_64-unknown-linux-gnu/test/run-make/fmt-write-bloat/fmt-write-bloat -L /git/rust/build/x86_64-unknown-linux-gnu/test/run-make/fmt-write-bloat/fmt-write-bloat  main.rs -O
nm /git/rust/build/x86_64-unknown-linux-gnu/test/run-make/fmt-write-bloat/fmt-write-bloat/main | "/git/rust/src/etc/cat-and-grep.sh" -v panicking panic_fmt panic_bounds_check pad_integral Display Debug
[[[ begin stdout ]]]
00000000000020b2 R anon.ad56e829de00bb5e05e79e94a78aaeca.0.llvm.11246502207772803501
0000000000004008 B __bss_start
0000000000004008 b completed.8060
                 w __cxa_finalize@@GLIBC_2.2.5
0000000000001070 t deregister_tm_clones
00000000000010e0 t __do_global_dtors_aux
0000000000003d20 d __do_global_dtors_aux_fini_array_entry
0000000000004000 D __dso_handle
0000000000003dd8 d _DYNAMIC
0000000000004008 D _edata
0000000000004010 B _end
0000000000001d6c T _fini
0000000000001120 t frame_dummy
0000000000003d18 d __frame_dummy_init_array_entry
0000000000002524 r __FRAME_END__
0000000000003f98 d _GLOBAL_OFFSET_TABLE_
                 w __gmon_start__
000000000000217c r __GNU_EH_FRAME_HDR
0000000000001000 t _init
0000000000003d20 d __init_array_end
0000000000003d18 d __init_array_start
                 w _ITM_deregisterTMCloneTable
                 w _ITM_registerTMCloneTable
0000000000001290 T __libc_csu_fini
0000000000001220 T __libc_csu_init
                 U __libc_start_main@@GLIBC_2.2.5
00000000000011c0 T main
00000000000010a0 t register_tm_clones
00000000000011b0 T rust_begin_unwind
0000000000001040 T _start
0000000000004008 D __TMC_END__
0000000000001d60 T _ZN36_$LT$T$u20$as$u20$core..any..Any$GT$7type_id17hec09389da73b5e7cE
0000000000001c30 T _ZN4core3fmt3num3imp52_$LT$impl$u20$core..fmt..Display$u20$for$u20$u64$GT$3fmt17h54377f781f7a5c8fE
0000000000001c30 T _ZN4core3fmt3num3imp54_$LT$impl$u20$core..fmt..Display$u20$for$u20$usize$GT$3fmt17h4fa1341a9dac6525E
00000000000012c0 T _ZN4core3fmt5write17h100dc89493224d93E
0000000000001a80 t _ZN4core3fmt9Formatter12pad_integral12write_prefix17h204c271bcefaaa8cE
00000000000015b0 T _ZN4core3fmt9Formatter12pad_integral17hb075f5592a5d8a70E
00000000000012a0 T _ZN4core3ops8function6FnOnce9call_once17hd1e4d320fcc785faE.llvm.14856878211942904523
0000000000001130 t _ZN4core3ptr13drop_in_place17h8d4e61ec59d31a0eE
0000000000001ad0 t _ZN4core3ptr13drop_in_place17hd5cc673701fb5dfbE
0000000000001ba0 t _ZN4core4iter8adapters3zip16Zip$LT$A$C$B$GT$3new17h5f4b1cffaf34d111E
0000000000001be0 t _ZN4core4iter8adapters3zip16Zip$LT$A$C$B$GT$3new17hf19b0aefdc63476aE
0000000000001ae0 T _ZN4core9panicking18panic_bounds_check17ha31d9d2fa83bb84dE
0000000000001b60 T _ZN4core9panicking9panic_fmt17he9332856f0d02b11E
0000000000001140 t _ZN50_$LT$$RF$mut$u20$W$u20$as$u20$core..fmt..Write$GT$10write_char17h3bc2d3f42fbd7666E
0000000000001150 t _ZN50_$LT$$RF$mut$u20$W$u20$as$u20$core..fmt..Write$GT$9write_fmt17h0fd2e1dfd6fb1e2aE
00000000000011a0 t _ZN50_$LT$$RF$mut$u20$W$u20$as$u20$core..fmt..Write$GT$9write_str17h46dc1af754709fa5E

[[[ end stdout ]]]
Error: should not match: panicking
Error: should not match: panic_fmt
Error: should not match: panic_bounds_check
Error: should not match: pad_integral
Error: should not match: Displayconsole
```
</details>

bors commented:
:pushpin: Commit 356d5b5b2addd6f34b0d6809635dd08a4712b126 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 356d5b5b2addd6f34b0d6809635dd08a4712b126 -->
<!-- homu: {"type":"Approved","sha":"356d5b5b2addd6f34b0d6809635dd08a4712b126","approver":"Mark-Simulacrum"} -->

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #78151

bors commented:
:pushpin: Commit 356d5b5b2addd6f34b0d6809635dd08a4712b126 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 356d5b5b2addd6f34b0d6809635dd08a4712b126 -->
<!-- homu: {"type":"Approved","sha":"356d5b5b2addd6f34b0d6809635dd08a4712b126","approver":"Mark-Simulacrum"} -->

jonas-schievink commented:
@bors r- rollup=iffy failed in https://github.com/rust-lang/rust/pull/78197#issuecomment-713852295

m-ou-se commented:
Moved the test from `run-make` to `run-make-fulldeps`. Not because it needs 'full deps', but because `run-make-fulldeps` tests are only run for the host and not for cross-compilation targets. (It'd be nice if the test also checked those targets, but it looks like the `run-make` infrastructure is not really prepared for that.)

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 484d9ebc293ae5e6011e7b7ee99a887d47f3f590 has been approved by `dtolnay`

<!-- @bors r=dtolnay 484d9ebc293ae5e6011e7b7ee99a887d47f3f590 -->
<!-- homu: {"type":"Approved","sha":"484d9ebc293ae5e6011e7b7ee99a887d47f3f590","approver":"dtolnay"} -->

bors commented:
:hourglass: Testing commit 484d9ebc293ae5e6011e7b7ee99a887d47f3f590 with merge 425b627f95a2426e71069b1e64b2d3bcb976571b...
<!-- homu: {"type":"BuildStarted","head_sha":"484d9ebc293ae5e6011e7b7ee99a887d47f3f590","merge_sha":"425b627f95a2426e71069b1e64b2d3bcb976571b"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1305738018)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1305738018","builder_name":"checks-actions"} -->

m-ou-se commented:
CI failure on x86_64-gnu:

```
[[[ begin stdout ]]]
<...>
00000000000012d0 t _ZN36_$LT$T$u20$as$u20$core..any..Any$GT$7type_id17h0b23518670c50e58E
0000000000001370 T _ZN4core3fmt5write17h8c604c4264281a93E
00000000000012a0 t _ZN4core3ops8function6FnOnce9call_once17ha870097096df370fE
00000000000012c0 t _ZN4core3ptr13drop_in_place17h00a330c7ba55f8f6E
0000000000001130 t _ZN4core3ptr13drop_in_place17h382ec7c12788ff84E
00000000000012e0 T _ZN4core9panicking5panic17hd33b8bec11e2d3e4E
0000000000001330 T _ZN4core9panicking9panic_fmt17h44ca6b2da9ba6541E
0000000000001140 t _ZN50_$LT$$RF$mut$u20$W$u20$as$u20$core..fmt..Write$GT$10write_char17h5bd44f5c2d70f37fE
0000000000001150 t _ZN50_$LT$$RF$mut$u20$W$u20$as$u20$core..fmt..Write$GT$9write_fmt17hcf969479a75c9baaE
00000000000011a0 t _ZN50_$LT$$RF$mut$u20$W$u20$as$u20$core..fmt..Write$GT$9write_str17hfbcc62d0cf8d18f4E
<...>
[[[ end stdout ]]]
Error: should not match: panicking
Error: should not match: panic_fmt
```

Looks like it mostly worked: there's no `panic_bounds_check`. But the other checks I hadded for `panicking::panic` and `panicking::panic_fmt` fail. Maybe something changed on the main branch that introduced a panic where there was none before. I'll try to reproduce it locally.

m-ou-se commented:
> Probably because the CI has debug assertions enabled for std.

Yup. Enabling std's debug assertions produces the same test failure. Removing the two `debug_assert!()`s this PR adds makes it pass again.

Not sure what to do with this though. I'd really like to test that no panic functions end up in the binary, but removing these `debug_assert!()`s is probably not the best idea.

Mark-Simulacrum commented:
It should be possible to disable the test if debug asserts are enabled (I forget the syntax or whether it's supported though). I can search for that during my next review pass if you set this to waiting on reviewer.

m-ou-se commented:
> It should be possible to disable the test if debug asserts are enabled

I'm afraid that'd mean it's simply always disabled. Seems like debug assertions are enabled for std on pretty much every platform in CI. I'm not sure if the others even run the run-make-fulldeps tests.

Mark-Simulacrum commented:
Ah, hm. Well -- at least currently, the llvm-8 builder does not have debug asserts enabled. I think this test can also be a regular (non-fulldeps) run-make test. Maybe we want to explicitly mark that builder as disabling debug assertions, editing the comment in https://github.com/rust-lang/rust/blob/b6ac411f45d38d867ce9f689bbd5c3e7456d0f65/src/ci/docker/host-x86_64/x86_64-gnu-llvm-8/Dockerfile#L61-L65? Obviously not full coverage, but for this I don't think we really need it.

For codegen and run-make tests (i.e., not full deps tests) it should also be relatively quick to build a non-debug-assert enabled std in rustbuild just for them, though that's really not supported right now so we'd need some fairly annoying option threading to make that happen.

I'd personally also be fine disabling std debug asserts on the nopt builders or something like that -- that would give us more coverage as well...

m-ou-se commented:
> I think this test can also be a regular (non-fulldeps) run-make test.

Yeah, it originally was a `run-make` test, but I changed it to `run-make-fulldeps` as a hack to avoid running it for cross compilation targets: https://github.com/rust-lang/rust/pull/78122#issuecomment-713906449

Mark-Simulacrum commented:
Hm, I think there's a only-hosts comment you can put to achieve the same thing.

kennytm commented:
@bors r-

Mark-Simulacrum commented:
@m-ou-se I believe this is waiting on you to implement the fix suggested in https://github.com/rust-lang/rust/pull/78122#issuecomment-716666900, which would permanently designate the llvm-8 (now 9, IIRC) builder to be debug-assert free and test this test only when debug assertions are off. Do you need something from me to coordinate that?

m-ou-se commented:
Oh, thanks for the reminder! I'll probably have time to do this in the next few days.

m-ou-se commented:
Looks like there's a lot more runners on which debug assertions are disabled than I originally thought. (i686-gnu, x86_64-gnu-llvm-8, x86_64-gnu-distcheck, i686-gnu-nopt, ..)

I've changed the test to only check for the `panicking` and `panic_fmt` symbols if NO_DEBUG_ASSERTIONS is set. If not set, it still checks for the other symbols.

> I think there's a only-hosts comment you can put to achieve the same thing.

I only found a `force-host`, but I'm not sure exactly what it does, or if it'd be the right thing.

I saw one other test that greps through `$(RUSTC) -vV` to determine the host. I've used that now to check for cross compiling, and moved it back to `run-make`.

Mark-Simulacrum commented:
This looks good to me. Could you squash the test adding commits? r=me with that done.

(FWIW I'm happy on smaller PRs to just always squash commits, and I think generally our policy of "add new commits" does more harm than good on average).

m-ou-se commented:
For most PRs I personally find it hard to review them with force-pushes between the reviews, which is why I usually use new commits after a review.

Squashed the test commits into one.

@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit 85b217d1d66d2cb61cf4b6d9cf5c0fc7a28c7831 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 85b217d1d66d2cb61cf4b6d9cf5c0fc7a28c7831 -->
<!-- homu: {"type":"Approved","sha":"85b217d1d66d2cb61cf4b6d9cf5c0fc7a28c7831","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit 85b217d1d66d2cb61cf4b6d9cf5c0fc7a28c7831 with merge 614ccbb44cca6b1bee828395ef4cebe629b27c92...
<!-- homu: {"type":"BuildStarted","head_sha":"85b217d1d66d2cb61cf4b6d9cf5c0fc7a28c7831","merge_sha":"614ccbb44cca6b1bee828395ef4cebe629b27c92"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1468856958)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1468856958","builder_name":"checks-actions"} -->

m-ou-se commented:
This failed on Windows because the no_std test (using `#[link(name = "c")]`) didn't link there. Added `# ignore-windows`.

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit 1da5780303b5df0505f20d7653778454225bb3cc has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 1da5780303b5df0505f20d7653778454225bb3cc -->
<!-- homu: {"type":"Approved","sha":"1da5780303b5df0505f20d7653778454225bb3cc","approver":"Mark-Simulacrum"} -->

sfackler commented on library/core/src/fmt/mod.rs:
```diff
@@ -1096,25 +1098,37 @@ pub fn write(output: &mut dyn Write, args: Arguments<'_>) -> Result {
 }
 
-fn run(fmt: &mut Formatter<'_>, arg: &rt::v1::Argument, args: &[ArgumentV1<'_>]) -> Result {
+unsafe fn run(fmt: &mut Formatter<'_>, arg: &rt::v1::Argument, args: &[ArgumentV1<'_>]) -> Result {
```
Nit: I'd add a `// SAFETY: arg and args must come from the same Arguments` comment to the newly unsafe functions as well to indicate the necessary preconditions when calling them.

bors commented:
:hourglass: Testing commit 1da5780303b5df0505f20d7653778454225bb3cc with merge cf9bfdb8726c07974c8db93f70ee7213ee20c563...
<!-- homu: {"type":"BuildStarted","head_sha":"1da5780303b5df0505f20d7653778454225bb3cc","merge_sha":"cf9bfdb8726c07974c8db93f70ee7213ee20c563"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1471511768)
Approved by: Mark-Simulacrum
Pushing cf9bfdb8726c07974c8db93f70ee7213ee20c563 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1471511768"},"merge_sha":"cf9bfdb8726c07974c8db93f70ee7213ee20c563"} -->
